File: leet/0858_mirror_reflection/0858_mirror_reflection.js
// LeetCode 858. Mirror Reflection
// O(p) time, O(1) space
/**
 * @param {number} p
 * @param {number} q
 * @return {number}
 */
var mirrorReflection = function(p, q) {
    let a = 1, b = 1;
    while (a * p != b * q)
        a = parseInt(++b * q / p);
    if (a % 2 == 1 && b % 2 == 0)
        return 2;
    if (a % 2 == 1 && b % 2 == 1)
        return 1;
    if (a % 2 == 0 && b % 2 == 1)
        return 0;
};

File: leet/0858_mirror_reflection/0858_mirror_reflection.c
// LeetCode 858. Mirror Reflection
// O(p) time, O(1) space

int mirrorReflection(int p, int q){
    int x = 1, y = 1;
    while (y * p != x * q)
        y = ++x * q / p;
    if (y % 2 == 0 && x % 2 == 1)
        return 0;
    if (y % 2 == 1 && x % 2 == 1)
        return 1;
    if (y % 2 == 1 && x % 2 == 0)
        return 2;
    return -1;
}
